From 8c8824fffa9cb52a1fa656bfb863c074bd2c7b7f Mon Sep 17 00:00:00 2001
From: Max Krummenacher <max.krummenacher@toradex.com>
Date: Fri, 16 May 2025 13:23:54 +0000
Subject: [PATCH] vvcam: dwe: platform_driver::remove: make compatible with
 newer and older kernels

Make the external kernel module compatible with kernels before 6.11 and
kernels 6.11 and later.
The commit 0edb555a65d1 ("platform: Make platform_driver::remove()
return void") added with kernel 6.11 changed the remove() function to
return void.

As the external kernel module might be built against older kernel
versions change the code to conditionally use the old or the new
remove() function pointer type.

Prevents build errors with the '.remove' API change.
| ../isp_driver_of.c:797:19: error: initialization of void (*)(struct platform_device *)
|    from incompatible pointer type int (*)(struct platform_device *) [-Werror=incompatible-pointer-types]
|      797 |         .remove = isp_hw_remove,

This completes the following commits which unconditionally use the new
remove() function pointer type.

commit e470f211909b ("LF-12820-1: vvcam: dwe: Fix build issue on rebase kernel next-20240618")
commit c2cf5b7d9c91 ("LF-12820-2: vvcam: isp: Fix build issue on rebase kernel next-20240618")
commit b5a4eaf38f49 ("LF-12820-3: vvcam: video: Fix build issue on rebase kernel next-20240618")

Upstream-Status: Inappropriate [oe specific]
Signed-off-by: Max Krummenacher <max.krummenacher@toradex.com>
---
 vvcam/v4l2/dwe/dwe_driver_of.c |  8 ++++++++
 vvcam/v4l2/isp/isp_driver_of.c | 16 ++++++++++++++++
 vvcam/v4l2/video/video.c       |  6 +++++-
 3 files changed, 29 insertions(+), 1 deletion(-)

diff --git a/vvcam/v4l2/dwe/dwe_driver_of.c b/vvcam/v4l2/dwe/dwe_driver_of.c
index 0901abf597b5..e555508273af 100644
--- a/vvcam/v4l2/dwe/dwe_driver_of.c
+++ b/vvcam/v4l2/dwe/dwe_driver_of.c
@@ -471,7 +471,11 @@ dewarp_destory_fake_pdev:
 	return rc;
 }
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 11, 0)
+static int dwe_hw_remove(struct platform_device *pdev)
+#else
 static void dwe_hw_remove(struct platform_device *pdev)
+#endif
 {
 	struct dwe_device *dwe_dev;
 	int dev_id;
@@ -491,7 +495,11 @@ static void dwe_hw_remove(struct platform_device *pdev)
 	}
 	dwe_fake_pdev_destory();
 	pr_info("vvcam dewarp driver removed\n");
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 11, 0)
+	return 0;
+#else
 	return;
+#endif
 }
 
 static int dwe_system_suspend(struct device *dev)
diff --git a/vvcam/v4l2/isp/isp_driver_of.c b/vvcam/v4l2/isp/isp_driver_of.c
index 0632f61319f3..e3ef3008c407 100644
--- a/vvcam/v4l2/isp/isp_driver_of.c
+++ b/vvcam/v4l2/isp/isp_driver_of.c
@@ -693,14 +693,22 @@ err_put_isp:
 	return rc;
 }
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 11, 0)
+static int isp_hw_remove(struct platform_device *pdev)
+#else
 static void isp_hw_remove(struct platform_device *pdev)
+#endif
 {
 	struct isp_device *isp = platform_get_drvdata(pdev);
 	int rc;
 
 	pr_info("enter %s\n", __func__);
 	if (!isp)
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 11, 0)
+		return -1;
+#else
 		return;
+#endif
 
 	tasklet_kill(&isp->ic_dev.tasklet);
 	vvbuf_ctx_deinit(&isp->bctx);
@@ -711,14 +719,22 @@ static void isp_hw_remove(struct platform_device *pdev)
 
 	if (rc) {
 		pr_err("vvcam isp detach pm domain failed\n");
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 11, 0)
+		return rc;
+#else
 		return;
+#endif
 	}
 
 	proc_remove(isp->pde);
 	pm_runtime_disable(&pdev->dev);
 	kfree(isp);
 	pr_info("vvcam isp driver removed\n");
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 11, 0)
+	return 0;
+#else
 	return;
+#endif
 }
 
 static int isp_system_suspend(struct device *dev)
diff --git a/vvcam/v4l2/video/video.c b/vvcam/v4l2/video/video.c
index f60798518ef3..1ac293e31e34 100644
--- a/vvcam/v4l2/video/video.c
+++ b/vvcam/v4l2/video/video.c
@@ -2161,7 +2161,11 @@ probe_end:
 	return rc;
 }
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 11, 0)
+static int viv_video_remove(struct platform_device *pdev)
+#else
 static void viv_video_remove(struct platform_device *pdev)
+#endif
 {
 	struct viv_video_device *vdev;
 	int i;
@@ -2191,7 +2195,7 @@ static void viv_video_remove(struct platform_device *pdev)
 
 	media_device_unregister(&mdev);
 	media_device_cleanup(&mdev);
-	return;
+	return 0;
 }
 
 static struct platform_driver viv_video_driver = {
-- 
2.42.0

